### PR TITLE
Add mount options to nfs-client-provisioner

### LINKED ISSUE #8016: add storageClass.mountOption as configurable value
**Is this a request for help?**:
no

FEATURE REQUEST
add storageclass.mountOptions as a configurable option

**Version of Helm and Kubernetes**:
2.9.1
1.11.2
on Rancher RKE provisioned but this shouldn't matter. I can use helm charts in here as well.

**Which chart**:
nfs-client-provisioner

**What happened**:
For my nfs export to work as a volume, I'm having to specify some mount options and I can't do those with the nfs-client-provisioner.

**What you expected to happen**:
be able to utilize storageClass mountOption so that volumes inherit them at mount time from the class.

**How to reproduce it** (as minimally and precisely as possible):
https://kubernetes.io/docs/concepts/storage/storage-classes/#the-storageclass-resource

> ..Persistent Volumes that are dynamically created by a storage class will have the mount options specified in the mountOptions field of the class.
If the volume plugin does not support mount options but mount options are specified, provisioning will fail. Mount options are not validated on either the class or PV, so mount of the PV will simply fail if one is invalid..." 

**Anything else we need to know**:
I don't think so

### PR DESCRIPTION
**Pull request by OlegPS**
#### What this PR does / why we need it:
Add mountOption to nfs-client-provisioner chart

#### Which issue this PR fixes
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
  - fixes #8016 

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://www.helm.sh/blog/helm-dco/index.html) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md

### CODE REVIEW

**Comment by davidkarlsen** [comment]
@verwilst PTAL

**Comment by verwilst** [comment]
I'm currently not getting the reasoning behind this addition. Why would you need a PVC to mount remote NFS shares? Could you please elaborate on this? :-)

**Comment by OlegPS** [comment]
Hi Bort! The reason is need for specifying the mount options, ex. `nfsvers`. Some NFS shares doesnt mount without it. Thanks!

**Comment by davidkarlsen** [comment]
But claims is something your applications should create, which will ask nfs-client-provisioner to provide it. Basically claims and volumes do not belong to this chart - but your applications chart, which will get it's claims fulfilled by using the storageclass this chart is providing

**Comment by OlegPS** [comment]
Moreover, the same options should go to the any provisioned volume or it will not mount. At the moment there is no way to specify mount options in the chart.

**Comment by treksler** [comment]
Something like this (a PVC and a PV for the nfs-client-root) is absolutely necessary for mountOptions to work. Even more importantly, this is required for the provisioner pod to even start, if your NFS server does not support the default version.

However,  the way this PR is done is very confusing, as evidenced by all the confused comments.

I have been working on this over at https://github.com/treksler/charts and have not submitted a PR yet, because there are several things to work out yet.

1st. I am not the least bit convinced this should be optional/configurable. i.e. it doesn't hurt to create the pv/pvc explicitly in all cases. Nobody on earth is going to understand the option "persistentVolume.create" and why it is required. It is not clear that option refers to the internal mechanism of mounting the root volume, so that the provisioner can create folders. But, whether that is clear is beside the point. Nobody should have to think about it or worry about it. It is an implementation detail. It allows the provisioner to work in all cases. see kubernetes-incubator/external-storage#1008

The only option that is really needed is mountOptions, but that belongs on storageClass, not on this persistentVolume thing in this PR

2nd. persistentVolume.capacity is an interesting one. I have not figured out what to do with his yet myself. I am not sure whether to specify minimum or maximum. Since the volume is used to create folders, I bet something like 10MB would be more than enough and it would not need to be configurable, but i have not tested it yet.

3rd. I wasn't sure if the pvc name (in the deployment) and the pv name in the pvc needs to be dynamic. looks like they are dynamic in this PR. This makes sense, because you might create multiple storage classes and they should each have their own root pvc and pv. I will change my code to reflect.

4th. passing the mount options as a range is interesting. I am not expertly familiar with yaml and I have not tested this myself. perhaps this will allow for the backslash to be omitted in front of the comma , when passing mount options

**Comment by OlegPS** [comment]
Hi Treksler! Thanks for useful comments!
1st. Looks like you are absolutely right here. PV just a side effect for mount options and definitely can be hidden. I will review that part.
2nd. AFAIK the provisioned does not checks the capacity and just set it as in incoming PVC, so we can use any value here?..
3rd. Yes, we can have several provisioners deployed.
4th. This method working in my environment. Will be nice if you can test this too.
I see that you take some code from the PR, so we can continue improving this or open new one from you. Thanks!

**Comment by OlegPS** [comment]
@treksler PTAL

**Comment by EronWright** [comment]
I like the option of creating a PVC when the provisioner is deployed, because in that context it is acting as an application and (as @davidkarlsen pointed out) applications are sometimes expected to create PVCs but aren't expected to know how the claim is satisfied.    

To clarify, imagine a 'chain' of provisioners working together:
```
App = "User Application (e.g. janusgraph)"
NCP = "NFS Client Provisioner"
FileStore = "gcp-filestore-csi-driver"

+-----+                +-----+              +-----------+
| App |                | NCP |              | FileStore |
+-----+                +-----+              +-----------+
   |                      |                       |
   |                      | PVC("nfs-root")       |
   |                      |---------------------->|
   |                      |                       |
   |                      |           PV("/root") |
   |                      |<----------------------|
   |                      |                       |
   | PVC("nfs-client")    |                       |
   |--------------------->|                       |
   |                      |                       |
   |      PV("/root/pv1") |                       |
   |<---------------------|                       |
   |                      |                       |
```

Some questions:
1. Why does this PR include a PV?
2. Are the mount options being introduced here related to the "nfs-client" class?
3. Does the PV that is generated by NCP automatically pick up the mount options from the associated storage class?
Thanks for the clarification.

**Comment by OlegPS** [comment]
Hi Eron!

> 1. Why does this PR include a PV?

Without PV, nfs-client-provisioner will not start, because claim for “nfs-root” will not be satisfied. So after “helm install” user should create corresponding PV manually. But I assume that after deployment better to have working provisioner without additional intervention and manual resource creation.

> 2. Are the mount options being introduced here related to the "nfs-client" class?

This is standard kubernetes PV mountOptions for NFS - https://kubernetes.io/docs/concepts/storage/persistent-volumes/#mount-options

> 3. Does the PV that is generated by NCP automatically pick up the mount options from the associated storage class?

Yes, specified options will be applied to “nfs-root” PV and to the created StorageClass. Any dynamic provisioned PV will inherits options from it’s class.

**Comment by EronWright** [comment]
Thanks @OlegPS for the clarifications.  Now I understand that the patch is trying to set both the `mountOptions` of the app's volume and the `mountOptions` of NCP's backend volume, with a single setting.

May I suggest that this PR be split in two:
1.  Add `.Values.storageClass.mountOptions` which simply configures the templated storage class.
2. Add `.Values.nfs.storageClass` which generates a PVC (for an external provider to satisfy, e.g. FileStore).  Also: `capacity`, `selector`.   Let the external provider deal with the backend's options.

**Comment by OlegPS** [comment]
Eron, I'm sorry, but i didn't get what "external provider" is.
nfs-client-provisioner is controller that dynamically creates PersistentVolume for incoming claim, based on the "already externally created" NFS share. It does not support any filestore provider, except single NFS share. If the share need the options to be mounted, we should specify them in all cases - mount to provisioner itself and mount to application claim. Thanks!

**Comment by EronWright** [comment]
Seems to me that the NFS Client Provisioner shouldn't be responsible for the mount options associated with its backend.   Mount options are usually conveyed in the `PersistentVolume`, and in this context the NCP is the _consumer_ of the backend PV.

Meanwhile, the NCP is also a _producer_ of a PV.  One could argue that the NCP should 'pick up' the mount options from the consumed PV to be propagated to the produced PV, and thus not require any configuration.   Or, one could argue that the mount options should be independent, e.g. to facilitate a security-related mount option.   

Basically this PR has chosen to commingle a bunch of independent options.

**Comment by treksler** [comment]
> Seems to me that the NFS Client Provisioner shouldn't be responsible for the mount options associated with its backend. Mount options are usually conveyed in the `PersistentVolume`, and in this context the NCP is the _consumer_ of the backend PV.

even before this PR, NCP already mounts the NFS store and consumes it. it just does so with default NFS client options (and without a PV, PVC)

> Meanwhile, the NCP is also a _producer_ of a PV. One could argue that the NCP should 'pick up' the mount options from the consumed PV to be propagated to the produced PV, and thus not require any configuration. 

it's the other way around. the user specifies the default mount options for produced PVs, and the consumed PV uses them as good enough defaults, in order to not require config.

> Or, one could argue that the mount options should be independent, e.g. to facilitate a security-related mount option.

like what? what is the use case?

> Basically this PR has chosen to commingle a bunch of independent options.
that was my suggestion. to separate them would make the user have to understand why two sets of mount options are needed, and i can't think of a single reason why anybody would want to do that. i.e. it's bad usability with no real tangible upside

in #8016, the main use-case was the need to switch away from the default nfs version, because the NFS server did not support it or because of some bug with the implementation. i.e. provide different defaults for the nfs client. 

perhaps the mountOptions should be at the top-level then to indicate that it is simply setting NFS client defaults
i.e. .Values.mountOptions

**Comment by OlegPS** [comment]
Hello! Let me clarify what and how we trying to achieve with this PR:
1. We have NFS share (nfs_server:/exported_dir) with non-standard mount options and it cannot be mounted without them.
2. We want to get working dynamic provision based on the given share with simple ```helm install```.

Also note:
- in order to provision Volumes, nfs-client-provisioner should mount NFS share to itself first;
- provisioned Volumes should use the same options to be mounted by application claim.

As is:
```
 ---------------------------
|          DEPLOYMENT       | 
| ...                       |
| volumes:                  |
|   - name: nfs-client-root |
|     nfs:                  |
|       server: nfs_server  |
|       path: /exported_dir |
 ---------------------------
```
In this case mount options cannot be specified, deployment/pod specification simply does not allow them. And Deployment will not start as it will not be able to mount our share without options. But only one place where we can specify options is PersistentVolume. So we will add it only for this purpose. Also we will hide it from users, it is just internal realization for mount options and there is no need for user to know about this.

To be:
```
 -------------------------------     -----------------------     ----------------------
|          DEPLOYMENT           |   |          PVC          |   |          PV           |
| ...                           |   | ...                   |   | ...                   |
| volumes:                      |   | storageClassName: ""  |   | nfs:                  |
|   - name: nfs-client-root     |   | selector:             |   |   server: nfs_server  |
|     persistentVolumeClaim:    |   |   matchLabels:        |   |   path: /exported_dir |
|       claimName: pvc-nfs-root |---|     name: pv-nfs-root |---| mountOptions: ...     |
 -------------------------------     -----------------------     -----------------------
```
Now user can just specify: ```nfs.server```, ```nfs.path```, ```nfs.mountOptions``` and get dynamic provision.

**Comment by OlegPS** [comment]
@davidkarlsen hi!

> But claims is something your applications should create, which will ask nfs-client-provisioner to provide it. 

Yes, but nfs-client-provisioner should mount root share first, to be able to provide any volumes.

> Basically claims and volumes do not belong to this chart - but your applications chart, which will get it's claims fulfilled by using the storageclass this chart is providing.

Right now the storageclass does not contain any mount options and suitable only for NFS share that can be mounted without them.

All explanation is also in discussion above. Do you need anything else from me, to be able to proceed with this PR in any way? Thank you!

**Comment by OlegPS** [comment]
Hi @verwilst 
Could you please review conversation above and help us with the PR. Thanks!

**Review by setonez — APPROVED** [review]
Tried in my environment and it works. It is properly mounting specified nfs.mountOptions for both, internal PV for provisioner and provisioned PVs.

**Comment by jdmaguire** [comment]
I've tried this and it works excellently (good job @OlegPS). It would be nice to get this in.

**Comment by miramar-labs** [comment]
could somebody please give an example of how to specify mount options (nfsvers=4) using --set ?

**Comment by OlegPS** [comment]
Hi! Sure, please take a look at 
https://github.com/helm/charts/issues/8016#issuecomment-428874486
